File: aeds02/tps/tp1/PalindromoRecursivo.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

#define MAXTAM 1000
#define true 1
#define false 0

int isFim(char* str);
char** getEntrada(char** entrada, int index);
void imprimir(char** saida, int index);
char** getPalindromo(char** entrada, int index);
int isPalindromo(char* entrada, int primeiro, int ultimo);

char** getPalindromo(char** entrada, int index) {
	if(isFim(entrada[index]) != true) {
		if(isPalindromo(entrada[index], 0, (int)strlen(entrada[index]) - 2) == true) {
			strcpy(entrada[index], "SIM"); 
		} else {
			strcpy(entrada[index], "NAO"); 
		}

		entrada = getPalindromo(entrada, index + 1);
	}
	return entrada;
}

int isPalindromo(char* entrada, int primeiro, int ultimo) {
	int resp = true;
	if(primeiro < (int)strlen(entrada) - 1) {
		if(entrada[primeiro] == entrada[ultimo]) {
			//printf("SIM: %c --- %c\n", entrada[primeiro], entrada[ultimo]);
			resp = isPalindromo(entrada, primeiro + 1, ultimo - 1);
		} else {
			//printf("NAO: %c --- %c\n", entrada[primeiro], entrada[ultimo]);
			resp = false;
			primeiro = (int)strlen(entrada);
		}
	}
	return resp;	
} 

char** getEntrada(char** entrada, int index) {
	entrada[index] = (char*) malloc(MAXTAM * sizeof(char));
	fgets(entrada[index], MAXTAM, stdin);
	if(isFim(entrada[index]) != true) {
		entrada = getEntrada(entrada, index + 1);
	}
	return entrada;
}

void imprimir(char** entrada, int index) {
	if(isFim(entrada[index]) != true) {
		printf("%s\n", entrada[index]);
		imprimir(entrada, index + 1);
	}
}

int isFim(char* str) {
	return (int)strlen(str) == 4 && str[0] == 'F' && str[1] == 'I' && str[2] == 'M' && str[3] == '\n';
}

int main() {
	char** entrada = (char**) malloc(MAXTAM * sizeof(char*));
	char** saida = (char**) malloc(MAXTAM * sizeof(char*));

	entrada = getEntrada(entrada, 0);
	saida = getPalindromo(entrada, 0);
	imprimir(saida, 0);

	return 0;
}

File: aeds02/tps/tp1/PalindromoRecursivo.java
class PalindromoRecursivo {
	public static String[] getPalindromo(String[] entrada) {
		return getPalindromo(entrada, 0);
	}
	public static String[] getPalindromo(String[] entrada, int index) {
		//String[] resp = new String[1000];

		if(!isFim(entrada[index])) {
			if(isPalindromo(entrada[index])) {
				entrada[index] = "SIM";
			} else {
				entrada[index] = "NAO";
			}

			entrada = getPalindromo(entrada, index + 1);
		}
		
		return entrada;
	}

	public static boolean isPalindromo(String str) {
		return isPalindromo(str, 0);
	}
	public static boolean isPalindromo(String str, int index) {
		boolean resp = true;
		int primeiro = index;
		int ultimo = (str.length() - 1) - index;

		if(index < str.length()) {
			if(str.charAt(primeiro) == str.charAt(ultimo)) {
				resp = isPalindromo(str, index + 1);
			} else {
				resp = false;
				index = str.length();
			}
		}

		return resp;
	}

	public static String[] getEntrada(String[] entrada) {
		return getEntrada(entrada, 0);
	}
	public static String[] getEntrada(String[] entrada, int numEntrada) {
		entrada[numEntrada] = MyIO.readLine();
		if(!isFim(entrada[numEntrada])) entrada = getEntrada(entrada, numEntrada + 1);
		return entrada;
	}

	public static void imprimir(String[] saida) {
		imprimir(saida, 0);
	}
	public static void imprimir(String[] saida, int numSaida) {
		if(!isFim(saida[numSaida])) {
			MyIO.println(saida[numSaida]);
			imprimir(saida, numSaida + 1);
		}
	}

	public static boolean isFim(String str) {
		return str.length() == 3 && str.charAt(0) == 'F' && str.charAt(1) == 'I' && str.charAt(2) == 'M';
	} 

	public static void main(String[] args) {
		String[] entrada = new String[1000];
		String[] saida = new String[1000];

		entrada = getEntrada(entrada);
		saida = getPalindromo(entrada);
		//MyIO.println(saida[0] + " " + saida[1]);
		imprimir(saida);
	}
}
